contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -374,6 +378,35 @@ tsql_precedence_info_t tsql_precedence_infos[] =
 
 #define TOTAL_TSQL_PRECEDENCE_COUNT (sizeof(tsql_precedence_infos)/sizeof(tsql_precedence_infos[0]))
 
+/* Following constants value are defined based on the special function list */
+#define SFUNC_MAX_ARGS 2			/* maximum number of args special function in special function list can have */
+#define SFUNC_MAX_VALID_TYPES 4		/* maximum number of valid types supported argument of function in special function list can have */
+
+/* struct to store details of valid types supported for a argument */
+typedef struct tsql_valid_arg_type
+{
+	int     len;                                      /* length of list of valid types for the argument */
+	char   *valid_types[SFUNC_MAX_VALID_TYPES];       /* list of valid type name supported for the argument */
+	Oid     valid_types_oid[SFUNC_MAX_VALID_TYPES];   /* list of valid type oid supported for the argument */
+} tsql_valid_arg_type_t;
+
+/* struct to store details of special function */
+typedef struct tsql_special_function
+{
+	const char             *nsp;                              /* namespace of special function */
+	const char             *funcname;                         /* name of special function */
+	const char             *formatted_funcname;				  /* formatted name of special function */
+	int                     nargs;                            /* number of arguments of special function */
+	tsql_valid_arg_type_t   valid_arg_types[SFUNC_MAX_ARGS];  /* list for storing details of all the valid types supported for each arguments */
+} tsql_special_function_t;
+
+tsql_special_function_t tsql_special_function_list[] = 
+{
+	{"sys", "trim", "Trim", 2, {{4, {"char","varchar","nchar","nvarchar"}, {InvalidOid, InvalidOid, InvalidOid, InvalidOid}}, {4, {"char","varchar","nchar","nvarchar"}, {InvalidOid, InvalidOid, InvalidOid, InvalidOid}}}}
```

Comment:
What about LTRIM/RTRIM/LEFT/RIGHT? Don't they need to be validated as well?

Reply:
String functions `LTRIM/RTRIM/LEFT/RIGHT` does not require any special handling, it can be implemented using sql functions. Validation of input argument of those functions are handled in their sql function definitions.